### ISSUE TITLE
Existential quantifier forces bound variable to appear in type?

### ISSUE THREAD

**Issue by martinberger** [rems-project/sail#415]
This program compiles, as it should:

    type exists_int = { 'n. int('n)}
    function f_good() -> exists_int = 1

However, if I replace the type `int('n)}` with `unit` the resulting program 

    type exists_unit = { 'n. unit}
    function f_bad() -> exists_unit = ()

I am not sure this is a bug. Please advise.

**Comment by Alasdair** [comment]
This isn't a bug I think, but the error isn't very helpful.

It'll infer that `()` has type `unit`, and then it needs to check that `unit` is a subtype of `{'n. unit}`. This will generate a constraint that has a universal quantifier (essentially we are trying to prove that unit is a subtype of unit for all values of `'n`). In this case this ends up being something like `(assert (forall ((n Int)) true))` which is trivial, but in general for decidability we don't allow constraints with universal quantifiers so it gets rejected.

I am not sure if there is any use for a type `{'n. T}` where `T` doesn't mention `'n`. I think the only reason we don't reject it as not well-formed is we allow types like `{'n 'm, 'n == 'm. int('m)}` as they can sometimes be generated by type inference.

**Comment by martinberger** [comment]
Agreed that such types are not particularly useful. It just arose during minimisation of the examples I cooked up for the other issues.

**Comment by martinberger** [comment]
Those edge cases could be mentioned in the documentation for existential types. ;-)

**Comment by martinberger** [comment]
BTW, in most logics, $\forall x. \phi$ and $\exists x. \phi$ are both equivalent to $\phi$ if $x$ does not occur freely in $\phi$. So why can we not 'just' rewrite `{'n. t}` to `t` when `'n` does not occur in `t`? Surely those two types are (semantically) isomorphic.
